### ISSUE TITLE
inner_join leaks memory

### ISSUE THREAD

**Issue by klmr** [hadley/dplyr#1295]
I `inner_join` large `data.frame`s (> 1.000.000 rows × 20.000 rows). Before the `inner_join`, my R process uses 600 MiB of memory. After the `inner_join`, the memory consumption has increased to 3.5 GiB. After running `gc()`, the memory usage once again drops down (to about 700 MiB).

I’d love to provide a reproducible example but I’m unsure how to produce sufficient amounts of example data. With less data, the problem also manifests but I’m not sure that R is expected to trigger a collection in these cases (since memory is obviously not scarce). However, with 3.5 GiB I’d expect that `inner_join` cleans up after itself, because that amount of memory *does* actually cause issues on my system (I’m assuming that the operating system starts swapping, and as a consequence the UI becomes unusably sluggish; this vanishes once a manual GC is triggered).

Apologies for the vague report. Maybe somebody else can corroborate this with better data.

**Comment by DavidArenburg** [comment]
If you are truly looking for a better idea, I would suggest to try `data.table`s binary join, which works best for inner joins. Some example [here](http://stackoverflow.com/questions/12773822/why-does-xy-join-of-data-tables-not-allow-a-full-outer-join-or-a-left-join) perhaps

**Comment by romainfrancois** [comment]
We just let R deal with garbage collection, and if it goes sufficiently down once you manually trigger it, it means we don't forget to release intermediate memory, so not much we can't do. I don't see dplyr starting  to make calls to `gc`.

**Comment by klmr** [comment]
Fair point. This seems more a bug in the R garbage collector …

**Comment by sgp667** [comment]
I encountered memory leaks as well.
I am using dplyr 0.5.0, R 3.3.0 , Win7 SP 1.

I noticed that memory leak occurs when I'm trying to join data frames using columns that have same name in both data frames, I was able to avoid it by using `by = c("A" = "B")` syntax.

I do understand that dplyr does not deal with garbage collection  but it seems that some bit of code in dplyr exposes this R bug and another bit of code that avoids this issue. 

Unfortunately I encountered this while working on proprietary data which I cannot share, and my attempts   to create sample data to reproduce this so far failed.
